### DISCUSSION TITLE
what is the format to write json file for question answering system???

### DISCUSSION THREAD

**Discussion by salama4ai** [facebookresearch/ParlAI#4166] [Q&A]
i need to write json file to use it in fine tuning question answering model, what is the json format for this task, where every conversation consists of only one question and only one answer, i.e no back and forth,exact one question and one answer, and then another completely different question and answer

assume i need to write this example:

text: where is the school ???                 label: go strait forward .
text: what is your name ???                  label: my name is jone .

**Comment by mojtaba-komeili** [comment]
If your entire dataset is as simple as your mentioned here (eg, "text: where is the school ??? label: go strait forward ."), you don't need a json file. You can simply use your text file, as we explained [here](https://parl.ai/docs/tutorial_task.html#quickstart-adding-a-new-dataset).

**Reply by salama4ai** [reply]
firstly thank you for your response i appreciate
ya, but when i try to use text directly i get an error tells me that the are issue in the text agent so i tried to use json to make sure from the sourse of the problem

**Reply by mojtaba-komeili** [reply]
Happy to help! Could you post the command that you are using and the stack trace of the error here please? That helps us to see what happens better.

**Reply by salama4ai** [reply]
**Logs**


```
/content/drive/MyDrive/polyencoder-master/ParlAI
05:19:50 | Opt:
05:19:50 |     allow_missing_init_opts: False
05:19:50 |     batchsize: 1
05:19:50 |     datapath: /usr/local/lib/python3.7/dist-packages/data
05:19:50 |     datatype: train:ordered
05:19:50 |     dict_class: None
05:19:50 |     display_add_fields: 
05:19:50 |     download_path: None
05:19:50 |     dynamic_batching: None
05:19:50 |     hide_labels: False
05:19:50 |     ignore_agent_reply: True
05:19:50 |     image_cropsize: 224
05:19:50 |     image_mode: raw
05:19:50 |     image_size: 256
05:19:50 |     init_model: None
05:19:50 |     init_opt: None
05:19:50 |     is_debug: False
05:19:50 |     jsonfile_datapath: /content/drive/MyDrive/polyencoder-master/ParlAI/data/f_train.json
05:19:50 |     jsonfile_datatype_extension: False
05:19:50 |     label_turns: secondspeaker
05:19:50 |     loglevel: info
05:19:50 |     max_display_len: 1000
05:19:50 |     model: None
05:19:50 |     model_file: None
05:19:50 |     multitask_weights: [1]
05:19:50 |     mutators: None
05:19:50 |     num_examples: 10
05:19:50 |     override: "{'task': 'jsonfile', 'jsonfile_datapath': '/content/drive/MyDrive/polyencoder-master/ParlAI/data/f_train.json'}"
05:19:50 |     parlai_home: /usr/local/lib/python3.7/dist-packages
05:19:50 |     starttime: Nov12_05-19
05:19:50 |     task: jsonfile
05:19:50 |     verbose: False
05:19:51 | Current internal commit: a6f368c9e27e3273db02d43b56d3216dad423573
05:19:51 | Current fb commit: a6f368c9e27e3273db02d43b56d3216dad423573
05:19:51 | creating task(s): jsonfile
05:19:51 | [loading data from json file into task:/content/drive/MyDrive/polyencoder-master/ParlAI/data/f_train.json]
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/worlds.py", line 1390, in _create_task_agents
    task_agents = my_module.create_agents(opt)  # type: ignore
AttributeError: module 'parlai.tasks.jsonfile.agents' has no attribute 'create_agents'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "parlai/scripts/display_data.py", line 114, in <module>
    DisplayData.main()
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/script.py", line 129, in main
    return cls._run_args(None)
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/script.py", line 101, in _run_args
    return cls._run_from_parser_and_opt(opt, parser)
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/script.py", line 108, in _run_from_parser_and_opt
    return script.run()
  File "parlai/scripts/display_data.py", line 109, in run
    return display_data(self.opt)
  File "parlai/scripts/display_data.py", line 71, in display_data
    world = create_task(opt, agent)
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/worlds.py", line 1441, in create_task
    world = create_task_world(opt, user_agents, default_world=default_world)
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/worlds.py", line 1405, in create_task_world
    task_agents = _create_task_agents(opt)
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/worlds.py", line 1393, in _create_task_agents
    return create_task_agent_from_taskname(opt)
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/teachers.py", line 2668, in create_task_agent_from_taskname
    task_agents = teacher_class(opt)
  File "/usr/local/lib/python3.7/dist-packages/parlai/tasks/jsonfile/agents.py", line 67, in __init__
    super().__init__(opt, shared)
  File "/usr/local/lib/python3.7/dist-packages/parlai/tasks/jsonfile/agents.py", line 58, in __init__
    self._setup_data(datafile)
  File "/usr/local/lib/python3.7/dist-packages/parlai/core/teachers.py", line 1698, in _setup_data
    conversations = Conversations(path)
  File "/usr/local/lib/python3.7/dist-packages/parlai/utils/conversations.py", line 199, in __init__
    self.conversations = self._load_conversations(datapath)
  File "/usr/local/lib/python3.7/dist-packages/parlai/utils/conversations.py", line 208, in _load_conversations
    f'Conversations at path {datapath} not found. '
```

**Reply by mojtaba-komeili** [reply]
Thanks, could you also add the ParlAI command you used that resulted in this output please?
According to the last error message it can not find your file. Are you sure you had the right path and you use it with the right flag?

**Reply by salama4ai** [reply]
`%cd {p}/ParlAI/`
`!python parlai/scripts/display_data.py --task jsonfile --jsonfile-datapath {p}/ParlAI/data/f_train.json`

**Reply by mojtaba-komeili** [reply]
Ok, I see what might be the problem now. Please have a look at the example json file [here](https://github.com/facebookresearch/ParlAI/blob/6380ad53ba74d88280a336ef5b74bce513fcdccf/parlai/tasks/jsonfile/example_data.json#L1). I just tried that with the following command and it did show me the conversations:
```
parlai display_data --task jsonfile --jsonfile-datapath {YOUR PARLAI PATH}/parlai/tasks/jsonfile/example_data.json
```

Either you do not have the correct path to the file, or your json file does not have the correct format. For the latter case you can fix it by copying the format from the sample file I linked above.

**Reply by salama4ai** [reply]
well, instead of the example json file you gave ?? 

what if i need to write different kind of conversation; i need to write conversation like that:
`{ "dialog": [ [`  
`{"id": "question", "text": "what is the color of the sky???"},    {"id": "answer", "text": "it's color is blue"},`
`{"id": "question", "text": "what is the expensive metal"},        {"id": "answer", "text": "gold is the expensive"}  ] ]}`
is the previous format is correct???

or do i need to write:-
`[{"id":1, "dialog": [ {"id": "question", "text": "what is the color of the sky???"},    {"id": "answer", "text": "it's color is blue"}  ]},`
`{"id":2, "dialog":  [{"id": "question", "text": "what is the expensive metal"},        {"id": "answer", "text": "gold is the expensive"}  ] }]`
ans so on

or what i need to write?? what is json format i need to write?
as the example you gave is only one dialogue what if  it's multiple separated dialogues as i need?

Thank you so much for your interesting for helping me

**Reply by mojtaba-komeili** [reply]
Considering the definition of the JsonTeacher ([link](https://github.com/facebookresearch/ParlAI/blob/main/parlai/tasks/jsonfile/agents.py)), I recommend you looking at the ConversationTeacher [here](https://github.com/facebookresearch/ParlAI/blob/6380ad53ba74d88280a336ef5b74bce513fcdccf/parlai/core/teachers.py#L1607), and its docstring for the expected format.
